stellargraph/layer/knowledge_graph.py

```diff
@@ -191,3 +192,135 @@ def build(self):
+            ``R^embedding_dimension`` is learnt for each node and each link type)
+
+        embeddings_initializer (str or func, optional): The initialiser to use for the embeddings.
+
+        embeddings_regularizer (str or func, optional): The regularizer to use for the embeddings.
+    """
+
+    def __init__(
+        self,
+        generator,
+        embedding_dimension,
+        embeddings_initializer="uniform",
+        embeddings_regularizer=None,
+    ):
+        if not isinstance(generator, KGTripleGenerator):
+            raise TypeError(
+                f"generator: expected KGTripleGenerator, found {type(generator).__name__}"
+            )
+
+        require_integer_in_range(embedding_dimension, "embedding_dimension", min_val=1)
+
+        graph = generator.G
+        self.num_nodes = graph.number_of_nodes()
+        self.num_edge_types = len(graph._edges.types)
+        self.embedding_dimension = embedding_dimension
+        self.embeddings_initializer = initializers.get(embeddings_initializer)
+        self.embeddings_regularizer = regularizers.get(embeddings_regularizer)
+
+    # layer names
+    _NODE = "DISTMULT_NODE"
+    _REL = "DISTMULT_EDGE_TYPE"
+
+    @staticmethod
+    def embeddings(model):
+        """
+        Retrieve the embeddings for nodes/entities and edge types/relations in the given model.
+
+        Args:
+            model (tensorflow.keras.Model): a Keras model created using a ``DistMult`` instance.
+
+        Returns:
+            A tuple of numpy arrays: the first element is the embeddings for nodes/entities
+            (``shape = number of nodes × k``), the second element is the embeddings for edge
+            types/relations (``shape = number of edge types x k``).
+        """
+        node = model.get_layer(DistMult._NODE).embeddings.numpy()
+        rel = model.get_layer(DistMult._REL).embeddings.numpy()
+
+        return node, rel
+
+    def _embed(self, count, name):
+        # FIXME(https://github.com/tensorflow/tensorflow/issues/33755): embeddings can't use
+        # constraints to be normalized: per section 4 in the paper, the embeddings should be
+        # normalised to have unit norm.
+        return Embedding(
+            count,
+            self.embedding_dimension,
+            name=name,
+            embeddings_initializer=self.embeddings_initializer,
+            embeddings_regularizer=self.embeddings_regularizer,
+        )
+
+    def __call__(self, x):
+        """
+        Apply embedding layers to the source, relation and object input "ilocs" (sequential integer
+        labels for the nodes and edge types).
+        """
+        e1_iloc, r_iloc, e2_iloc = x
+
+        # DistMult generates embeddings in R
+        node_embeddings = self._embed(self.num_nodes, self._NODE)
+        edge_type_embeddings = self._embed(self.num_edge_types, self._REL)
+
+        e1 = node_embeddings(e1_iloc)
+        r = edge_type_embeddings(r_iloc)
+        e2 = node_embeddings(e2_iloc)
+
```

Comment:
Shouldn't the above be called y_e1, y_e2, m_r (M_r?) since the tensors hold the node and relationship embeddings as per equation 2 in the paper?